### PR TITLE
Equality with conversions

### LINKED ISSUE #8855: (1.0 . to BigDecimal) != 1.0 - e.g. equality isn't working correctly
The support for non-builtin number types (as introduced by #8779) **doesn't work well with `==`**. As result of that certain [tests had to be disabled](https://github.com/enso-org/enso/pull/8779/commits/6b9bdca0c7854ecd5ded86b4f3501805ce54efe0). This is a follow up issue that tracks further work on the reimplementation of `EqualsNode`.

The initial idea:
- when comparing two objects of different type, check possible conversions first
- only then use comparators

Catch:
- warn when there is not a consistency of `hash` code!

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Fixes #8855 by providing support for `==` between different types with conversions.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The [documentation has been updated](https://github.com/enso-org/enso/blob/5d52d518fe18dd6ed150d444d0f1225e8e462358/docs/syntax/functions.md#custom-equality), if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides. 
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] Verify [benchmark results](https://github.com/enso-org/enso/actions/runs/7913598314), [2nd run](https://github.com/enso-org/enso/actions/runs/7928027568)
  - [x] Enforce consistency of `==` and `hash` code: ecd4b90b834766f33a27d6d721ab1d1173feaa80

### CODE REVIEW

File: test/Base_Tests/src/Data/Numbers_Spec.enso [hunk 1]
```diff
@@ -591,21 +587,25 @@ add_specs suite_builder =
...
             Ordering.compare v1 v2 . catch Incomparable_Values (_->42) . should_equal 42
 
-        group_builder.specify "Equality of complex and number" pending=Complex.pending_equality <|
+        group_builder.specify "Equality of complex and number" <|
             v1 = (Complex.new 3)
             v2 = 3
+            r1 = v1==v2
+            r2 = v2==v1
+            r1 . should_be_true
+            r2 . should_be_true
```
Comment (JaroslavTulach) [inline test/Base_Tests/src/Data/Numbers_Spec.enso:598]: @GregoryTravis, these tests are now passing with:
```
sbt:enso> runEngineDistribution --run test/Base_Tests Number
Number Conversions: [9/9, 130ms]
    - Complex plus Integer [24ms]
    - Integer plus Complex [12ms]
    - Equality of complex and complex [9ms]
    - Inequality of complex and complex [21ms]
    - Equality of complex and number [8ms]
    - Equality of number and complex [7ms]
    - Greater or equal of complex and complex [26ms]
    - Greater or equal of complex and number [10ms]
    - Greater or equal of number and complex [9ms]
```
e.g. functionally wise this PR offers enough for your to have `Big_Decimal` equality with other number types. Feel free to test it.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java [hunk 2]
```diff
@@ -40,366 +37,170 @@
...
+   *
+   * @param frame the stack frame we are executing at
+   * @param self the self object
+   * @param other the other object
+   * @return {@code true} if {@code self} and {@code that} seem equal
+   */
+  public boolean execute(
+      VirtualFrame frame, @AcceptsError Object self, @AcceptsError Object other) {
+    var areEqual = node.execute(frame, self, other);
+    if (!areEqual) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:88]: @Akirathan, I have renamed original `EqualsNode` to `EqualsSimpleNode`. Then I placed this new `EqualsNode` in front of the original one and in case the original one _reports inequality_ I try harder with conversions via `WithConversionNode`.

Functionally it should be fine. The effect on benchmarks needs yet to be verified.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAtomNode.java [hunk 3]
```diff
@@ -32,7 +34,7 @@ public static EqualsAtomNode build() {
     return EqualsAtomNodeGen.create();
   }
 
-  public abstract boolean execute(Atom left, Atom right);
+  public abstract boolean execute(VirtualFrame frame, Atom left, Atom right);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAtomNode.java:37]: Alas, @Akirathan, I had to add `frame` argument to all `Equals*Node`s. The possible conversion require context of `VirtualFrame`. As such the diff is pretty huge...

File: distribution/lib/Standard/Test/0.0.0-dev/src/Extensions.enso [hunk 4]
```diff
@@ -296,7 +296,7 @@ Number.should_equal : Float -> Float -> Integer -> Spec_Result
 Number.should_equal self that epsilon=0 frames_to_skip=0 =
     matches = case that of
         _ : Number -> self.equals that epsilon
-        _ -> False
+        _ -> self==that
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Test/0.0.0-dev/src/Extensions.enso:299]: 9f4cde3d9f620d31820f736a6796d4e9990ea168 - otherwise `Number` couldn't be `should_equal` to `Complex`.

**Comment by JaroslavTulach** [comment]
There are currently failures in [EnsoMultiValue conversions](https://github.com/enso-org/enso/actions/runs/7913585673/job/21601546343?pr=9070#step:8:5597). Looks like the tests need to be updated to the new behavior: bbbbf73

**Comment by JaroslavTulach** [comment]
First [benchmark results](https://github.com/enso-org/enso/actions/runs/7913598314) are here and at least the `equalsPrimitives` benchmark seems negatively affected:

equalsPrimitives

File: test/Base_Tests/src/Data/Numbers_Spec.enso (hunk 1 above)
Comment (GregoryTravis) [reply test/Base_Tests/src/Data/Numbers_Spec.enso:598]: What is necessary for this to work for a new type? Does it require a `from` conversion from `Number`, and a `Comparator`?

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java [hunk 5]
```diff
@@ -40,367 +37,184 @@
...
+              + self
+              + "[#"
+              + Long.toHexString(selfHash)
+              + "] got converted to "
+              + converted
+              + "[#"
+              + Long.toHexString(convertedHash)
+              + "]";
+      if (!ok) {
+        var ctx = EnsoContext.get(this);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:287]: As a result of this `-ea` only check, I had to update `Complex_Comparator` to return the same hash for numbers without imaginary part as the real part has. I also had to add comparator+hash code for `Fool`: ecd4b90

File: test/Base_Tests/src/Data/Numbers_Spec.enso (hunk 1 above)
Comment (JaroslavTulach) [reply test/Base_Tests/src/Data/Numbers_Spec.enso:598]: Yes, it requires `from` conversion (not necessarily from `Number`, possibly just from `Integer`) & comparator. Both as [provided by complex number](https://github.com/enso-org/enso/pull/9070/files#diff-d92d98dfa353a1da9423b90033f0370b50ddc904d299d134b132fe05630865b2R30) sample. Moreover it is essential to have compatible [hash code implementation](https://github.com/enso-org/enso/pull/9070#discussion_r1491152185) between original & converted values.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:287]: ### Flaw in the Design

I am afraid there is a flaw in the design. Once we define `Text.from (that:Any)  = that.to_text`, do we really want
```ruby
obj == obj.to_text
```
? As that's what we would get if we just try to apply conversion to match both sides of `==`.

f33faad mitigates the problem a bit by doing the conversion only when original and converted object have the **same hash code**. That fixes the most apparent problem, but still one could find some `Number` and a `Text` that accidentally have the same hashcode when number is `.to_text` converted...

### TL;DR

Keeping consistency of `==` and `hash` is important for our `Map` implementation. Doing these conversions is important for `Number` and `Decimal` equality. Doing these conversions for any `.from` conversion may yield surprising results.
  
### What to do with that?

Solution follows Greg's comment...

Comment (GregoryTravis) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:287]: Conversions from `Any` are very useful for end-user code, but perhaps they could be omitted from the set of `from` candidates when doing `==`? In the case of `==` we want to be more selective, and feel that we have precisely defined which conversions to do, without surprises.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:287]: ### Explicitly Define Both Comparators

To guarantee that the implementation of `==` is supposed to use `from` conversion we need the `Complex` (or `Rational` or `Decimal`) module to explicitly state that it supports equality for `Number` (or `Integer`). To do so, let's request the module to define **conversions of both types** to `Comparator`. E.g.:
```ruby
Comparable.from (_:Complex) = Complex_Comparator
Comparable.from (_:Number) = Complex_Comparator
```
Only when **both conversions are defined** and yield the same comparator (in the scope of the module), then `==` functionality will perform the conversion.

##### Rationale

By explicitly specifying multiple `Comparable.from` conversions, the module without any doubt declares its willingness to support the `==` and `hash` code operations for combination of such types. Detecting presence of such conversions is _sufficient check_ for performing the conversions. Conversions defined for other purpose `Text.from (that:Complex)` aren't going to be (accidentally) taken into account without appropriate `Comparable.from (_:Text)` definition.

Such **conversions are local**, however. Other modules like `Map` or `Set` will continue to use the original definition of `Comparable.from (_:Number)` coming from the `Numbers` module. The infrastructure will check for [consistency of hash](https://github.com/enso-org/enso/commit/ecd4b90b834766f33a27d6d721ab1d1173feaa80#diff-b4b39451e4a4157a026bd98878cf0a23c55bbc5245a37a21efcf2de676f59738R199) and yield an `AssertionError` when the `hash` consistency is violated.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:287]: Initial implementation is in a5c0674c3b

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java [hunk 6]
```diff
@@ -40,367 +40,223 @@
...
+          "thatType != null",
+          "selfType == findType(typeOfNode, self)",
+          "thatType == findType(typeOfNode, that)"
+        })
+    final boolean doConversionCached(
+        VirtualFrame frame,
+        Object self,
+        Object that,
+        @Shared("typeOf") @Cached TypeOfNode typeOfNode,
+        @CachedLibrary(limit = "3") TypesLibrary types,
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java]: Remove unused argument

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java [hunk 7]
```diff
@@ -40,367 +40,223 @@
...
+      var state = State.create(ctx);
+      try {
+        var thatAsSelf = convertNode.execute(convert, state, new Object[] {selfType, that});
+        var result = equalityNode.execute(frame, self, thatAsSelf);
+        assert !result || assertHashCodeIsTheSame(self, thatAsSelf);
+        return result;
+      } catch (ArityException ex) {
+        var assertsOn = false;
+        assert assertsOn = true;
+        if (assertsOn) {
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:260]: Why not just panic here? `ArityException` is unexpected here, no? So it should result in some IllegalStateException

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+import org.enso.interpreter.runtime.data.EnsoMultiValue;
+import org.enso.interpreter.runtime.data.atom.Atom;
+import org.enso.interpreter.runtime.data.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+import org.enso.polyglot.common_utils.Core_Text_Utils;
+
+@GenerateUncached
+public abstract class EqualsSimpleNode extends Node {
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java:26]: Minor: Rename to `EqualsPrimitiveNode`?

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsComplexNode.java [hunk 9]
```diff
@@ -39,24 +39,26 @@ public static EqualsComplexNode build() {
     return EqualsComplexNodeGen.create();
   }
 
-  public abstract boolean execute(@AcceptsError Object left, @AcceptsError Object right);
```
Comment (Akirathan) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsComplexNode.java:42]: Why was the `@AcceptsError` removed here?

**Comment by JaroslavTulach** [comment]
[Second run](https://github.com/enso-org/enso/actions/runs/7928027568) of benchmarks seem to indicate `equalsPrimitives` are now fine:

equalsPrimitives

**Review by Akirathan — APPROVED** [review]
LGTM. Please, don't forget to provide some docs somewhere how the equality works now. It's starting to get really complex.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsComplexNode.java (hunk 9 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsComplexNode.java:42]: `@AcceptError` is Enso Builtins DSL, but this node is no longer processed by the DSL. `EqualsNode` is and it does contain these annotations.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+  }
+
+  @Specialization
+  boolean equalsBoolInterop(
+      boolean self,
+      Object other,
+      @Shared("interop") @CachedLibrary(limit = "10") InteropLibrary interop) {
+    try {
+      return self == interop.asBoolean(other);
+    } catch (UnsupportedMessageException ex) {
```
Comment (4e6) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java:66]: Would it be more efficient to do the `isBoolean` check instead of recovering from the exception?

**Review by 4e6 — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java (hunk 10 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java:66]: Unlikely. The `UnsupportedMessageException` has overriden [fillInStackTrace()](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropException.html#fillInStackTrace--) - e.g. throwing and catching it is fast.

Moreover it is expected that code of `interop.asBoolean` is going to be inlined into the same compilation unit when emitting assembly code. Then throwing/catching works just like a `goto`.

Calling `isBoolean`would in fact brought in more bytecode as the catch would still need to be done (it is a checked exception). Probably there would be no difference in performance (as the inlining and elimination of duplicated code phases would do its job) and the generated assembly would be almost the same.

`try/catch` seems easier to type and clearly calls less than `isBoolean` check would.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java (hunk 8 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsSimpleNode.java:26]: The node also deals with non-primitive values (which it delegates to `EqualsComplexNode` for example). Current name is probably not perfect, but suggestion renaming would be misleading.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java (hunk 7 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:260]: The answer is probably the same as in [the previous discussion](https://github.com/enso-org/enso/pull/8779/files?file-filters%5B%5D=.java&show-deleted-files=true&show-viewed-files=true#r1467896387) about the topic.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java (hunk 6 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java]: Good catch. Removed.

**Comment by JaroslavTulach** [comment]
Running benchmarks again for [engine](https://github.com/enso-org/enso/actions/runs/7960186596) and [stdlib](https://github.com/enso-org/enso/actions/runs/7960191238). The results aren't bad, so let's integrate this.

**Review by GregoryTravis — APPROVED** [review]
